### ISSUE TITLE
ssa_rtol in SSAFD should scale as function of number of grid points

### ISSUE THREAD

**Issue by bueler** [pism/pism#136]
The following is from an email exchange AA <--> EB on 12 Jan 2013.  It should not be lost.

Ed, Constantine--

looks like a change from the default `ssa_relative_convergence` tolerance from 1e-4 to 1e-5 makes quite a difference: the large daily(?) fluctuations disappear using 1e-5. The plot shows cbase.

Andy

Andy & Constantine:

Cool.  That is the clearest piece of data on a hard issue which I have
ever seen.

Is the following thinking reasonable?:

The SSA iteration converges in SSAFD if the L^1 norm of the effective viscosity nuH is not changing much in a relative sense, i.e. if

```
  ||nuH_k+1 - nuH_k||_1  <  1e-4 * ||nuH_k||_1
```

But `||.||_1` is the discrete L^1 norm which acts on  v \in R^N  by

```
||v||_1 = \sum_{k=1}^N |v_k|
```

and which is not area weighted; for instance it is not the weighted
L^1 norm which is the average

```
||v||_average = (1/N) \sum_{k=1}^N |v_k|.
```

Thus I claim the convergence criterion we are using is effectively
weaker as the number of grid points gets larger.

As a concrete demonstration, in the following Matlab session I have
used the discrete 1 norm in the "same way" we use it for convergence.
In this demo think of the correspondence:
  N = Mx My
  v = nuH_k
  w = nuH_k+1
I repeat the random test four times for each N:

```
>> N=10; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  7.1775e-04
>> N=10; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  0.010945
>> N=10; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  0.0043261
>> N=10; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  0.0050450

>> N=1000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  4.5431e-05
>> N=1000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  4.3587e-05
>> N=1000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  7.1925e-05
>> N=1000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  8.5481e-05

>> N=100000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  1.3933e-06
>> N=100000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  4.6298e-07
>> N=100000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  7.3082e-07
>> N=100000; v=randn(N,1); w=v; w(1)=w(1)+0.1*randn(1); norm(v-w,1)/norm(v,1)
ans =  1.1707e-06
```

To relate this demonstration to the SSA solver, suppose there is only
one "difficult" grid location.  Between one SSA outer iteration and
the next, suppose that only that one location has a change in nuH, and
that change is of order 10% (as in the random perturbation of w(1)
above).  Our current convergence criterion is like

```
norm(v-w,1)/norm(v,1) < 1e-4
```

Then for N=10 grid points we would generally not pass the criterion
but for N = 1000 = 10^3 grid points we would (barely) and for N =
100000 = 10^5 grid points we would (easily).  But the one "difficult"
spot would be just as bad in each case.

Do you think I have the facts of the current convergence scheme correct?

Thus, should we have the following scaling in the SSA convergence tolerance?:

```
  ||nuH_k+1 - nuH_k||_1  <  ssa_rtol * Mx * My * ||nuH_k||_1
```

In this case the default "ssa_rtol" should change, however.  I suppose
most of the testing of the SSA we have done uses N = Mx \* My of size
10^4, i.e. grids of size 100 \* 100.  Perhaps what we want is

```
  ||nuH_k+1 - nuH_k||_1  <  ssa_rtol * (Mx * My / 10^4) * ||nuH_k||_1
```

and then use the same default for ssa_rtol=10^-4?  Or perhaps even
this modified scaling, again sticking to ssa_rtol=1e-4?:

```
  ||nuH_k+1 - nuH_k||_1  <  ssa_rtol * max(1, Mx * My / 10^4) * ||nuH_k||_1
```

This would mean that the new behavior was the same as the old behavior
for runs with N=Mx*My <= 10^4 but for "high res" runs the new scaling
would kick in.

Note that even the following elaboration might make sense because the
number of "problem" grid points should also increase with N though
perhaps not proportionally to it:

```
  ||nuH_k+1 - nuH_k||_1  <  ssa_rtol * max(1, sqrt(Mx * My) / 10^2) * ||nuH_k||_1
```

(Again with the same ssa_rtol=1e-4.) This last scaling would make
sense if the set of problem points was not zero dimensional (a set
number of fixed locations are problematical as the grid is refined)
but one-dimensional (a curve of points along the boundary of something
are problematical as the grid is refined).

Comments appreciated.

Ed

**Comment by bueler** [comment]
This is ancient but it seems like there might be a worthwhile idea here?

**Comment by ckhroulev** [comment]
@bueler Hmm. This idea is closely linked to the odd convergence criterion used by SSAFD. I should create a new version of SSAFD using SNES to manage Picard iterations instead. I think I should be able to re-use most of the code, e.g. the matrix assembly. What do you think?

**Comment by bueler** [comment]
Sure.  A caution is that you will probably need to use https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/SNES/SNESSetConvergenceTest.html and write your own norm if you want to use the L^1 criterion applied on the effective viscosity.  (At least PETSc has an L^1 norm.  I think that's a fancy new PETSc feature! :smiley:)

**Comment by bueler** [comment]
To clarify, my comment assumes you would try both the old criterion and a sensible new criterion once you've moved over to SNES.

**Comment by ckhroulev** [comment]
Right. I suppose I should implement the old criterion so that we can compare it to the one using a norm of the residual... but aside from that the idea is to *drop* the old convergence criterion. :smile:
